File: transcoder-service/storage.py
import os
import boto3
from botocore.client import Config as BotoConfig
from config import Config

class StorageManager:
    def __init__(self):
        self.s3 = boto3.resource(
            's3',
            endpoint_url=Config.MINIO_ENDPOINT,
            aws_access_key_id=Config.MINIO_ACCESS_KEY,
            aws_secret_access_key=Config.MINIO_SECRET_KEY,
            config=BotoConfig(signature_version='s3v4'),
            region_name='us-east-1'
        )
        self.raw_bucket_name = Config.MINIO_RAW_BUCKET
        self.processed_bucket_name = Config.MINIO_PROCESSED_BUCKET

    def download_video(self, s3_path, local_path):
        """Downloads an object from the raw bucket to a local file."""
        print(f"Downloading {s3_path} to {local_path} from {self.raw_bucket_name}...")
        self.s3.Bucket(self.raw_bucket_name).download_file(s3_path, local_path)
        return local_path

    def upload_directory(self, local_dir, remote_dir):
        """Uploads all files in a directory to the processed bucket."""
        print(f"Uploading {local_dir} to {remote_dir} in {self.processed_bucket_name}...")
        bucket = self.s3.Bucket(self.processed_bucket_name)
        
        for root, dirs, files in os.walk(local_dir):
            for file in files:
                local_file_path = os.path.join(root, file)
                rel_path = os.path.relpath(local_file_path, local_dir)
                remote_path = os.path.join(remote_dir, rel_path)
                
                bucket.upload_file(local_file_path, remote_path)
        
        return os.path.join(remote_dir, "master.m3u8")

File: transcoder-service-go/internal/storage/storage.go
package storage

import (
	"context"
	"os"
	"path/filepath"
	"transcoder-service-go/internal/config"

	"github.com/minio/minio-go/v7"
	"github.com/minio/minio-go/v7/pkg/credentials"
)

type StorageManager struct {
	client                     *minio.Client
	RawBucket, ProcessedBucket string
}

func NewStorageManager(cfg *config.Config) (*StorageManager, error) {
	endpoint := cfg.MinioEndpoint
	secure := false

	if len(endpoint) > 8 && endpoint[:8] == "https://" {
		endpoint = endpoint[8:]
		secure = true
	} else if len(endpoint) > 7 && endpoint[:7] == "http://" {
		endpoint = endpoint[7:]
		secure = false
	}

	client, err := minio.New(endpoint, &minio.Options{
		Creds:  credentials.NewStaticV4(cfg.MinioAccessKey, cfg.MinioSecretKey, ""),
		Secure: secure,
	})
	return &StorageManager{client: client, RawBucket: cfg.MinioRawBucket, ProcessedBucket: cfg.MinioProcessedBucket}, err
}

func (m *StorageManager) DownloadVideo(ctx context.Context, s3Path, localPath string) error {
	return m.client.FGetObject(ctx, m.RawBucket, s3Path, localPath, minio.GetObjectOptions{})
}

func (m *StorageManager) UploadDirectory(ctx context.Context, localDir, remoteDir string) (string, error) {
	err := filepath.Walk(localDir, func(path string, info os.FileInfo, err error) error {
		if err != nil || info.IsDir() {
			return err
		}
		relPath, _ := filepath.Rel(localDir, path)
		_, err = m.client.FPutObject(ctx, m.ProcessedBucket, filepath.Join(remoteDir, relPath), path, minio.PutObjectOptions{})
		return err
	})
	return filepath.Join(remoteDir, "master.m3u8"), err
}
